File: packages/proxy/src/providers.ts
import { BUILTIN_PROVIDER_ENDPOINTS, type BuiltinProviderEndpoint } from "./constants.js";
import type {
  NormalizedProxyProvider,
  ProxyModelAlias,
  ProxyModelAliases,
  ProxyProviderInterface,
  ProxyProviderOptions,
} from "./types.js";

interface ProviderDefinition {
  providerInterface: ProxyProviderInterface;
  requiresBaseUrl: boolean;
}

const PROVIDER_DEFINITIONS: Record<string, ProviderDefinition> = {
  amazon_bedrock: {
    providerInterface: "amazon_bedrock",
    requiresBaseUrl: true,
  },
  anthropic: {
    providerInterface: "anthropic",
    requiresBaseUrl: false,
  },
  azure_openai: {
    providerInterface: "azure_openai",
    requiresBaseUrl: true,
  },
  deepseek: {
    providerInterface: "deepseek",
    requiresBaseUrl: false,
  },
  gemini: {
    providerInterface: "gemini",
    requiresBaseUrl: false,
  },
  groq: {
    providerInterface: "groq",
    requiresBaseUrl: false,
  },
  mistral: {
    providerInterface: "mistral",
    requiresBaseUrl: false,
  },
  moonshotai: {
    providerInterface: "moonshotai",
    requiresBaseUrl: false,
  },
  ollama: {
    providerInterface: "ollama",
    requiresBaseUrl: true,
  },
  openai: {
    providerInterface: "openai",
    requiresBaseUrl: false,
  },
  plano: {
    providerInterface: "plano",
    requiresBaseUrl: true,
  },
  qwen: {
    providerInterface: "qwen",
    requiresBaseUrl: true,
  },
  together_ai: {
    providerInterface: "together_ai",
    requiresBaseUrl: false,
  },
  xai: {
    providerInterface: "xai",
    requiresBaseUrl: false,
  },
  zhipu: {
    providerInterface: "zhipu",
    requiresBaseUrl: false,
  },
};

function parseProviderAndModel(input: ProxyProviderOptions): { model: string; provider: string } {
  if (input.provider) {
    return {
      model: input.model,
      provider: input.provider.toLowerCase(),
    };
  }

  const [provider, ...modelParts] = input.model.split("/");
  if (!provider || modelParts.length === 0) {
    throw new Error(
      `Provider is required for model "${input.model}". Use { provider, model } or a "<provider>/<model>" model string.`,
    );
  }

  return {
    model: modelParts.join("/"),
    provider: provider.toLowerCase(),
  };
}

function sanitizeBaseUrlPathPrefix(pathname: string) {
  if (!pathname || pathname === "/") {
    return undefined;
  }

  return pathname.endsWith("/") ? pathname.slice(0, -1) : pathname;
}

function hashSuffix(input: string) {
  let hash = 2166136261;

  for (const character of input) {
    hash ^= character.codePointAt(0) ?? 0;
    hash = Math.imul(hash, 16777619);
  }

  return (hash >>> 0).toString(16).padStart(8, "0").slice(0, 8);
}

function createClusterName(
  providerInterface: ProxyProviderInterface,
  hostname: string,
  port: number,
  pathPrefix?: string,
) {
  const safeBase = `${providerInterface}_${hostname}_${port}`.replaceAll(/[^a-zA-Z0-9_]/g, "_");
  const suffix = hashSuffix(`${hostname}:${port}${pathPrefix ?? ""}`);

  return `${safeBase}_${suffix}`;
}

function normalizeOneProvider(input: ProxyProviderOptions): NormalizedProxyProvider {
  const { model, provider } = parseProviderAndModel(input);
  const definition = PROVIDER_DEFINITIONS[provider];

  if (definition && input.providerInterface) {
    if (input.providerInterface !== definition.providerInterface) {
      throw new Error(
        `Provider "${provider}" already maps to providerInterface "${definition.providerInterface}", but "${input.providerInterface}" was supplied.`,
      );
    }
  }

  const providerInterface = input.providerInterface ?? definition?.providerInterface;

  if (!providerInterface) {
    throw new Error(`Custom provider "${provider}" requires providerInterface to be set.`);
  }

  if (definition?.requiresBaseUrl && !input.baseUrl) {
    throw new Error(`Provider "${provider}" requires baseUrl to be set in native proxy mode.`);
  }

  if (!definition && !input.baseUrl) {
    throw new Error(`Custom provider "${provider}" must provide baseUrl so Envoy can route to it.`);
  }

  let endpointHost: string | undefined;
  let endpointPort: number | undefined;
  let endpointProtocol: "http" | "https" | undefined;
  let baseUrlPathPrefix: string | undefined;
  let clusterName: string | undefined;

  if (input.baseUrl) {
    const parsed = new URL(input.baseUrl);
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
      throw new Error(`baseUrl for provider "${provider}" must use http or https.`);
    }

    endpointHost = parsed.hostname;
    endpointProtocol = parsed.protocol === "https:" ? "https" : "http";
    endpointPort =
      parsed.port === "" ? (endpointProtocol === "https" ? 443 : 80) : Number(parsed.port);
    baseUrlPathPrefix = sanitizeBaseUrlPathPrefix(parsed.pathname);
    clusterName = createClusterName(
      providerInterface,
      endpointHost,
      endpointPort,
      baseUrlPathPrefix,
    );
  }

  return {
    accessKey: input.apiKey,
    baseUrl: input.baseUrl,
    baseUrlPathPrefix,
    clusterName,
    default: input.default,
    endpointHost,
    endpointPort,
    endpointProtocol,
    model,
    name: input.name ?? `${provider}/${model}`,
    passthroughAuth: input.passthroughAuth,
    provider,
    providerInterface,
  };
}

export function normalizeProviders(inputs: ProxyProviderOptions[]) {
  if (inputs.length === 0) {
    throw new Error("At least one provider must be configured.");
  }

  const normalized = inputs.map(normalizeOneProvider);
  const seenNames = new Set<string>();
  let defaultCount = 0;

  for (const provider of normalized) {
    if (seenNames.has(provider.name)) {
      throw new Error(`Provider name "${provider.name}" must be unique.`);
    }
    seenNames.add(provider.name);

    if (provider.default) {
      defaultCount += 1;
    }
  }

  if (defaultCount > 1) {
    throw new Error("Only one provider can be marked as default.");
  }

  return normalized;
}

export function normalizeModelAliases(aliases?: ProxyModelAliases) {
  const normalized: Record<string, ProxyModelAlias> = {};
  if (!aliases) {
    return normalized;
  }

  for (const [name, alias] of Object.entries(aliases)) {
    normalized[name] =
      typeof alias === "string"
        ? {
            target: alias,
          }
        : alias;
  }

  return normalized;
}

export function getBuiltinProviderEndpoint(
  providerInterface: ProxyProviderInterface,
): BuiltinProviderEndpoint | undefined {
  return BUILTIN_PROVIDER_ENDPOINTS[providerInterface as keyof typeof BUILTIN_PROVIDER_ENDPOINTS];
}

File: internal/gateway/providers.go
package gateway

import (
	"fmt"
	"hash/fnv"
	"net/url"
	"strings"

	"github.com/akarachen/proxy-up/internal/config"
)

// providerRequiresBaseURL indicates whether a built-in provider interface requires a base URL.
var providerRequiresBaseURL = map[string]bool{
	"amazon_bedrock": true,
	"anthropic":      false,
	"azure_openai":   true,
	"deepseek":       false,
	"gemini":         false,
	"groq":           false,
	"mistral":        false,
	"moonshotai":     false,
	"ollama":         true,
	"openai":         false,
	"plano":          true,
	"qwen":           true,
	"together_ai":    false,
	"xai":            false,
	"zhipu":          false,
}

// builtinEndpoint is the upstream endpoint for a provider that has a known base URL.
type builtinEndpoint struct {
	host     string
	port     int
	protocol string // "http" or "https"
}

var builtinEndpoints = map[string]builtinEndpoint{
	"anthropic":   {host: "api.anthropic.com", port: 443, protocol: "https"},
	"deepseek":    {host: "api.deepseek.com", port: 443, protocol: "https"},
	"gemini":      {host: "generativelanguage.googleapis.com", port: 443, protocol: "https"},
	"groq":        {host: "api.groq.com", port: 443, protocol: "https"},
	"mistral":     {host: "api.mistral.ai", port: 443, protocol: "https"},
	"moonshotai":  {host: "api.moonshot.ai", port: 443, protocol: "https"},
	"openai":      {host: "api.openai.com", port: 443, protocol: "https"},
	"together_ai": {host: "api.together.xyz", port: 443, protocol: "https"},
	"xai":         {host: "api.x.ai", port: 443, protocol: "https"},
	"zhipu":       {host: "open.bigmodel.cn", port: 443, protocol: "https"},
}

// NormalizedProvider is a provider with all fields resolved and validated.
type NormalizedProvider struct {
	AccessKey         string
	BaseURL           string
	BaseURLPathPrefix string
	ClusterName       string
	Default           bool
	EndpointHost      string
	EndpointPort      int
	EndpointProtocol  string // "http" or "https"
	Model             string
	Name              string
	PassthroughAuth   bool
	Provider          string
	ProviderInterface string
}

// GetClusterName returns the cluster name for routing purposes.
// If ClusterName is explicitly set, it uses that; otherwise it falls back
// to ProviderInterface.
func (p NormalizedProvider) GetClusterName() string {
	if p.ClusterName != "" {
		return p.ClusterName
	}
	return p.ProviderInterface
}

// hashSuffix returns the first 8 hex digits of the FNV-1a 32-bit hash of input.
// Matches the TypeScript implementation used for cluster name generation.
func hashSuffix(input string) string {
	h := fnv.New32a()
	h.Write([]byte(input))
	return fmt.Sprintf("%08x", h.Sum32())
}

func createClusterName(providerInterface, hostname string, port int, pathPrefix string) string {
	raw := fmt.Sprintf("%s_%s_%d", providerInterface, hostname, port)
	safe := strings.Map(func(r rune) rune {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') || r == '_' {
			return r
		}
		return '_'
	}, raw)
	suffix := hashSuffix(fmt.Sprintf("%s:%d%s", hostname, port, pathPrefix))
	return fmt.Sprintf("%s_%s", safe, suffix)
}

func sanitizePathPrefix(pathname string) string {
	if pathname == "" || pathname == "/" {
		return ""
	}
	return strings.TrimSuffix(pathname, "/")
}

func normalizeProvider(input config.ProviderOptions) (NormalizedProvider, error) {
	provider := strings.ToLower(input.Provider)
	model := input.Model

	// Parse "provider/model" shorthand if provider field is empty.
	if provider == "" {
		parts := strings.SplitN(model, "/", 2)
		if len(parts) < 2 || parts[0] == "" || parts[1] == "" {
			return NormalizedProvider{}, fmt.Errorf(
				"provider is required for model %q — use the provider field or \"provider/model\" format", model)
		}
		provider = strings.ToLower(parts[0])
		model = parts[1]
	}

	providerInterface := input.ProviderInterface
	if providerInterface == "" {
		providerInterface = provider
	}

	requiresBase, known := providerRequiresBaseURL[providerInterface]
	if !known && input.BaseURL == "" {
		return NormalizedProvider{}, fmt.Errorf("custom provider %q must set baseUrl", provider)
	}
	if known && requiresBase && input.BaseURL == "" {
		return NormalizedProvider{}, fmt.Errorf("provider %q requires baseUrl", provider)
	}

	norm := NormalizedProvider{
		AccessKey:         input.APIKey,
		BaseURL:           input.BaseURL,
		Default:           input.Default,
		Model:             model,
		Name:              input.Name,
		PassthroughAuth:   input.PassthroughAuth,
		Provider:          provider,
		ProviderInterface: providerInterface,
	}
	if norm.Name == "" {
		norm.Name = fmt.Sprintf("%s/%s", provider, model)
	}

	if input.BaseURL != "" {
		u, err := url.Parse(input.BaseURL)
		if err != nil {
			return NormalizedProvider{}, fmt.Errorf("invalid baseUrl for %q: %w", provider, err)
		}
		if u.Scheme != "http" && u.Scheme != "https" {
			return NormalizedProvider{}, fmt.Errorf("baseUrl for %q must use http or https", provider)
		}

		norm.EndpointHost = u.Hostname()
		norm.EndpointProtocol = u.Scheme

		if u.Port() == "" {
			if u.Scheme == "https" {
				norm.EndpointPort = 443
			} else {
				norm.EndpointPort = 80
			}
		} else {
			fmt.Sscanf(u.Port(), "%d", &norm.EndpointPort)
		}

		norm.BaseURLPathPrefix = sanitizePathPrefix(u.Path)
		norm.ClusterName = createClusterName(providerInterface, norm.EndpointHost, norm.EndpointPort, norm.BaseURLPathPrefix)
	}

	return norm, nil
}

// NormalizeProviders validates and normalizes the provider list.
func NormalizeProviders(inputs []config.ProviderOptions) ([]NormalizedProvider, error) {
	if len(inputs) == 0 {
		return nil, fmt.Errorf("at least one provider must be configured")
	}

	out := make([]NormalizedProvider, 0, len(inputs))
	seen := map[string]bool{}
	defaults := 0

	for _, in := range inputs {
		norm, err := normalizeProvider(in)
		if err != nil {
			return nil, err
		}
		if seen[norm.Name] {
			return nil, fmt.Errorf("provider name %q must be unique", norm.Name)
		}
		seen[norm.Name] = true
		if norm.Default {
			defaults++
		}
		out = append(out, norm)
	}

	if defaults > 1 {
		return nil, fmt.Errorf("only one provider can be marked as default")
	}
	return out, nil
}

// NormalizeModelAliases normalizes the model aliases map.
func NormalizeModelAliases(aliases map[string]config.ModelAlias) map[string]string {
	result := make(map[string]string, len(aliases))
	for k, v := range aliases {
		result[k] = v.Target
	}
	return result
}
